lib/links/parser.rb

```diff
@@ -2,51 +2,58 @@
 
 class Parser
 
-    def initialize(url)
-        @elements = {a: ["href"],
-                img: ["src"],
-                script: ["src"],
-                frame: ["src"],
-                iframe: ["src"],
-                link: ["href"],
-                source: ["src", "srcset"],
-                track: ["src"] }
-
-        @url = url
+  def initialize(url)
+    @elements = { a: ['href'],
+                  img: ['src'],
+                  script: ['src'],
+                  frame: ['src'],
+                  iframe: ['src'],
+                  link: ['href'],
+                  source: ['src', 'srcset'],
+                  track: ['src'] }
+
+    @url = url
+  end
+
+  def run
+    all_links = []
+    @doc = Nokogiri::HTML(HTTParty.get(@url))
+
+    @elements.each do |elem, attr_list|
+      elems = get_elems(elem)
+
+      attr_list.each do |attr|
+        links = extract_links(elems, attr)
+        all_links = (all_links << links).flatten
```

Comment:
I wonder if this could be done with [`Enumerable#inject`](https://ruby-doc.org/core-2.4.0/Enumerable.html#method-i-inject)?